Addition of Continuous Integration (CI) Benchmark Script Along with Data Generation Mechanism .

RudraCodesForU opened the pull request:
### Your checklist for this pull request
Please review the [guidelines for contributing](CONTRIBUTING.md) to this repository.

- [x] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). Don't request your master!
- [x] Make sure you are making a pull request against the **dev branch** (left side). Also you should start *your branch* off *our dev*.
- [x] Check the commit's or even all commits' message styles matches our requested structure.

### Issue number(s) that this pull request fixes
- Fixes #2150

### List of changes to the codebase in this pull request
- Added the Benchmark Script for CITests and data generating mechanism(DGM).
- Added documentation for easy implementation in docs/Benchmark_for_CITests.md
- Sample Output is given below.

![Screenshot 2025-06-01 212041](https://github.com/user-attachments/assets/d16ee007-05f5-4e79-99a4-4cb6651d4071)

RudraCodesForU committed:
Added data mechanisms for benchmarking

RudraCodesForU committed:
Updates in DGM

RudraCodesForU committed:
Updated the DGMs and benchmark scripts added

RudraCodesForU committed:
Updated the Benchmark-Script

RudraCodesForU committed:
Updated the BenchmarkScript along with the Documentation

RudraCodesForU committed:
Deleted previous results

RudraCodesForU committed:
Merge branch 'dev' into feature/ci-benchmark-script

RudraCodesForU commented:
@ankurankan , I have made the PR regarding Issue #2150 , Kindly pls review it.

ankurankan commented on pgmpy/benchmark/data_generator_mechanism.py:
```diff
@@ -0,0 +1,174 @@
+        X = a1 * Z + e1
+        Y = a2 * Z + e2
+        e1, e2 ~ N(0, noise_std^2)
+    Parameters:
```
We use numpydoc style docstring. Have a look at any other function/class in the package for examples.

ankurankan commented on pgmpy/benchmark/data_generator_mechanism.py:
```diff
@@ -0,0 +1,174 @@
+    return pd.DataFrame(data)
+
+
+def nonlinear_gaussian(n_samples=1000, effect_size=1.0, noise_std=1.0, seed=None):
```
We need the option to specify multiple conditional variables here as well.

ankurankan replied:
What does non-linear gaussian mean here?

ankurankan commented on pgmpy/benchmark/data_generator_mechanism.py:
```diff
@@ -0,0 +1,174 @@
+    e1 = rng.normal(scale=noise_std, size=n_samples)
+    e2 = rng.normal(scale=noise_std, size=n_samples)
+    X = effect_size * Z.sum(axis=1) + e1
+    Y = effect_size * Z.sum(axis=1) + e2
```
This only generates conditionally independent data. We will need to generate conditionally dependent data as well to test power.

ankurankan commented on pgmpy/benchmark/data_generator_mechanism.py:
```diff
@@ -0,0 +1,174 @@
+#  NEW DGMS WILL BE ADDED HERE
+
+
+def user_defined(generator_func, **kwargs):
```
No need for this.

ankurankan commented on pgmpy/benchmark/data_generator_mechanism.py:
```diff
@@ -0,0 +1,174 @@
+    "user_defined": user_defined,
+}
+
+# Example usage for testing:
```
No need for this.

ankurankan commented on pgmpy/benchmark/CI_benchmark.py:
```diff
@@ -0,0 +1,164 @@
+    print(f"Running DGM: {dgm_name}")
+    compatible_tests = DGM_TO_CITESTS[dgm_name]
+    for n in sample_sizes:
+        for noise in noise_levels:
```
No need to vary both noise and effect sizes

ankurankan commented on pgmpy/benchmark/CI_benchmark.py:
```diff
@@ -0,0 +1,164 @@
+                    # Ensure categorical columns are marked as such for discrete/mixed DGMs
+                    if dgm_name in ["discrete_categorical", "mixed_data"]:
+                        for col in df.columns:
+                            if df[col].dtype != float and df[col].nunique() < 10:
```
The data generating function should specify the type of variable in each column

ankurankan commented on docs/Benchmark_for_CITests.md:
```diff
@@ -0,0 +1,148 @@
+# Conditional Independence (CI) Tests Benchmark Documentation
```
This should show the plots generated from the benchmarks. I think we should focus first on getting the benchmarks done and then we can work on how to present it in the documentation.

ankurankan commented on pgmpy/benchmark/data_generator_mechanism.py:
```diff
@@ -0,0 +1,174 @@
+    Z = rng.normal(size=n_samples)
+    e1 = rng.normal(scale=noise_std, size=n_samples)
+    e2 = rng.normal(scale=noise_std, size=n_samples)
+    X = np.sin(effect_size * Z) + e1
```
Does the "effect_size" parameter here behave like an effect size? I don't think increasing or decreasing it has any effect on the strength of their relationship.

ankurankan reviewed (changes requested):
@RudraCodesForU Thanks for the PR. I have added a few comments. Broadly:

1. We should try to keep the data generating mehods arguments as similar as possible.
2. We should define a function that runs the benchmarks. This function should accept a list of data generating functions and CI tests that we should run the benchmark for.
3. We should add some plotting code so that we can visualize the results.

RudraCodesForU committed:
Refactor the benchmark Script along with plotting added and Also made the dgms modular

RudraCodesForU committed:
Merge branch 'feature/ci-benchmark-script' of https://github.com/RudraCodesForU/pgmpy into feature/ci-benchmark-script

RudraCodesForU committed:
Updated the Benchmark along with plotting and DGMs modulation

RudraCodesForU committed:
Merge branch 'dev' into feature/ci-benchmark-script

RudraCodesForU commented:
Hello @ankurankan , I have addressed the changes you requested above. Also added the plotting part. Refactored the Benchmarking and made every dgms similar. Kindly review it.

ankurankan commented:
@RudraCodesForU I think the plots need some work. I would suggest trying to recreate Figure 2 and 3 from https://arxiv.org/pdf/2206.04356 for the different data generating process and tests.

RudraCodesForU commented:
> @RudraCodesForU I think the plots need some work. I would suggest trying to recreate Figure 2 and 3 from https://arxiv.org/pdf/2206.04356 for the different data generating process and tests.

Thanks for the advice, looking into it asap.

RudraCodesForU committed:
Updated the Benchmark as per needed

RudraCodesForU committed:
Merge branch 'dev' into feature/ci-benchmark-script

RudraCodesForU commented:
Hello @ankurankan , Sorry for the delay, plots with diff conditional variables and for diff sample sizes have been performed. I tried to replicate the plots given in https://arxiv.org/pdf/2206.04356  as much by varying the variables, size of samples. PLs review the changes.

ankurankan commented on pgmpy/benchmark/CI_benchmark.py:
```diff
@@ -0,0 +1,271 @@
+
+
+def run_benchmark(
+    dgms,
```
Please add the default values here instead.

RudraCodesForU replied:
All the changes, performed

ankurankan commented on pgmpy/benchmark/CI_benchmark.py:
```diff
@@ -0,0 +1,271 @@
+):
+    results = []
+    for dgm_name, dgm in dgms.items():
+        print(f"\nRunning DGM: {dgm_name}")
```
I don't think we need printing. Try to add tqdm progress bars instead.

ankurankan commented on pgmpy/benchmark/CI_benchmark.py:
```diff
@@ -0,0 +1,271 @@
+        for n_cond_vars in n_cond_vars_list:
+            for n in sample_sizes:
+                for noise in noise_levels:
+                    for significance_level in significance_levels:
```
We don't need to iterate over `significance_level`. We should be getting the p values from the tests and then use that to construct the plots.

ankurankan commented on pgmpy/benchmark/CI_benchmark.py:
```diff
@@ -0,0 +1,271 @@
+            for n in sample_sizes:
+                for noise in noise_levels:
+                    for significance_level in significance_levels:
+                        for effect_type, eff in [("null", 0.0), ("alt", 1.0)]:
```
We should have more variation in effect size as this is what would tell us sensitivity of the tests. This should also be an argument.

ankurankan replied:
We also don't need the null and alt here. You can separate this loop into conditionally independent and dependent cases. And for the alt case we will need to vary the effect size more.

ankurankan commented on pgmpy/benchmark/CI_benchmark.py:
```diff
@@ -0,0 +1,271 @@
+        compatible_tests = dgm_to_citests[dgm_name]
+        for n_cond_vars in n_cond_vars_list:
+            for n in sample_sizes:
+                for noise in noise_levels:
```
No need to do both noise_levels and effect size. I would suggest just keeping the effect size.

ankurankan commented on pgmpy/benchmark/CI_benchmark.py:
```diff
@@ -0,0 +1,271 @@
+                                            "Y",
+                                            z_cols,
+                                            df,
+                                            boolean=True,
```
We should be getting the p-values here and use that to construct our type 1 error plots.

ankurankan commented on pgmpy/benchmark/CI_benchmark.py:
```diff
@@ -0,0 +1,271 @@
+                                for test_name in compatible_tests:
+                                    ci_func = ci_tests[test_name]
+                                    accepted = None
+                                    try:
```
Why do we need a try-except block here? Shouldn't we already know which tests would work for the given data generating mechanism?

RudraCodesForU committed:
Updated the benchmark as per required

RudraCodesForU commented:
@ankurankan, I have made the change, you requested, kindly look into it.

ankurankan reviewed (changes requested):
@RudraCodesForU Thanks for the changes, I think it's starting to get there. The way I am envisioning this at the moment is that we build a website around this where a user can first selects the type of data (categorical, continuous), then the DGM,  and based on that we show them the comparison plots. This would require us to write the plotting functionality in javascript. I would ideally like to add similar benchmarks for other classes of algorithms such as causal discovery, pairwise orientation, causal identification. 

I also think it would be best that we create a separate repo for this and the website. I have just created this repo: https://github.com/pgmpy/algo-benchmarks. Could you please open this PR there and let me know if you would be interested in working on the website part of this?

RudraCodesForU committed:
Merge branch 'dev' into feature/ci-benchmark-script

RudraCodesForU commented:
> @RudraCodesForU Thanks for the changes, I think it's starting to get there. The way I am envisioning this at the moment is that we build a website around this where a user can first selects the type of data (categorical, continuous), then the DGM, and based on that we show them the comparison plots. This would require us to write the plotting functionality in javascript. I would ideally like to add similar benchmarks for other classes of algorithms such as causal discovery, pairwise orientation, causal identification.
> 
> I also think it would be best that we create a separate repo for this and the website. I have just created this repo: https://github.com/pgmpy/algo-benchmarks. Could you please open this PR there and let me know if you would be interested in working on the website part of this?

Thank you @ankurankan! I’m definitely interested in working on the website part for deploying the benchmarks. I’ll move this PR to the https://github.com/pgmpy/algo-benchmarks repo as suggested and will keep you updated once it’s up. Looking forward to collaborating further on both the benchmarking scripts and the website!

RudraCodesForU commented:
@ankurankan, I had made a PR on PGMPY/ALGO-BENCHMARK. Kindly pls look into it

ankurankan commented:
Ref: https://github.com/pgmpy/algo-benchmarks/pull/1
